### PR TITLE
[FLINK-11822][table-planner-blink] Introduce Flink metadata handlers

### PR DESCRIPTION
**Pull request by godfreyhe**
## What is the purpose of the change

*Introduce Flink metadata handlers*

## Brief change log

  - *adds Flink metadata handlers, including `FlinkRelMdPercentageOriginalRow`, `FlinkRelMdNonCumulativeCost`, `FlinkRelMdCumulativeCost`, `FlinkRelMdRowCount`, `FlinkRelMdSize`, `FlinkRelMdSelectivity`, `FlinkRelMdDistinctRowCoun`, `FlinkRelMdPopulationSize`, `FlinkRelMdColumnUniqueness`, `FlinkRelMdUniqueKeys`*
  - *adds dependent classes  for test: `FlinkCalciteCatalogReader`, `FlinkRelBuilder`, `FlinkRelOptClusterFactory`, `FlinkTypeSystem` and `FlinkRelOptTable`*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added test for each metadata handler*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/util/FlinkRelMdUtil.scala [hunk 1]
```diff
@@ -18,30 +18,326 @@
...
+import java.math.BigDecimal
+import java.util
 
 import scala.collection.JavaConversions._
+import scala.collection.mutable
 
 /**
   * FlinkRelMdUtil provides utility methods used by the metadata provider methods.
   */
 object FlinkRelMdUtil {
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/util/FlinkRelMdUtil.scala]: This utility is confusing with `FlinkRelOptUtil`, i don't see a clear response boundary between these two utilities after you adding more methods to this class.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/util/NodeResourceUtil.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+import org.apache.flink.table.api.TableConfigOptions;
+
+import static org.apache.flink.configuration.ConfigOptions.key;
+
+/**
+ * Deal with resource config for {@link org.apache.flink.table.plan.nodes.exec.ExecNode}.
+ */
+public class NodeResourceUtil {
+
+	public static final ConfigOption<Integer> SQL_RESOURCE_INFER_OPERATOR_PARALLELISM_MIN =
```
Comment (KurtYoung) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/util/NodeResourceUtil.java]: I don't think we should place config name in an utility class

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/metadata/AggCallSelectivityEstimator.scala [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+  *
+  * A filter predicate on an Aggregate may contain two parts:
+  * one is on group by columns, another is on aggregate call's result.
+  * The first part is handled by [[SelectivityEstimator]],
+  * the second part is handled by this Estimator.
+  *
+  * @param agg aggregate node
+  * @param mq  Metadata query
+  */
+class AggCallSelectivityEstimator(agg: RelNode, mq: FlinkRelMetadataQuery)
```
[the remark this answers is not in the document]
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/metadata/AggCallSelectivityEstimator.scala]: `SelectivityEstimator` uses general algorithm, and it does not care about the predicate is on an agg-call or not. However, `AggCallSelectivityEstimator` only estimates predicate on an agg-call. So there are many difference in their logic.

the comments in `AggCallSelectivityEstimator` could be updated: `Estimates selectivity of rows meeting an agg-call predicate on an Aggregate.`

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/logical/FlinkLogicalAggregate.scala [hunk 4]
```diff
@@ -161,7 +162,14 @@ object FlinkLogicalAggregate {
       groupSets: util.List[ImmutableBitSet],
       aggCalls: util.List[AggregateCall]): FlinkLogicalAggregate = {
     val cluster = input.getCluster
-    val traitSet = cluster.traitSet().replace(FlinkConventions.LOGICAL).simplify()
-    new FlinkLogicalAggregate(cluster, traitSet, input, indicator, groupSet, groupSets, aggCalls)
+    val traitSet = cluster.traitSetOf(Convention.NONE)
+    //FIXME: FlinkRelMdDistribution requires the current RelNode to compute
```
[the remark this answers is not in the document]
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/logical/FlinkLogicalAggregate.scala]: Those `FIXME`s are not added in this PR, they are added when `SplitAggregateRule` and `FlinkLogicalRelFactories` are introduced in Blink.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/plan/metadata/FlinkMetadata.java [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+		interface Handler extends MetadataHandler<FlinkDistribution> {
+			FlinkRelDistribution flinkDistribution(RelNode r, RelMetadataQuery mq);
+		}
+	}
+
+	/**
+	 * Metadata about the modified property of a RelNode. For example, an aggregate RelNode
+	 * contains a max aggregate function whose result value maybe modified increasing.
+	 */
+	public interface ModifiedMonotonicity extends Metadata {
```
[the remark this answers is not in the document]
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/plan/metadata/FlinkMetadata.java]: It means the result can be updatable on some keys. The original name is `RowUpdateMonotonicity` in Blink, and then it was rename to `ModifiedMonotonicity`.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/util/FlinkRelMdUtil.scala (hunk 1 above)
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/util/FlinkRelMdUtil.scala]: `FlinkRelMdUtil` provides utility methods(which are not defined in Calcite `RelMdUtil`) used by the metadata provider methods. (`Md` means metadata) 
`FlinkRelOptUtil` provides utility methods (which are not defined in Calcite `RelOptUtil`) used for use in optimizing RelNodes.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/util/NodeResourceUtil.java (hunk 2 above)
Comment (godfreyhe) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/util/NodeResourceUtil.java]: `NodeResourceConfig` may be better than `NodeResourceUtil`.

**Pull request merged**